Add ephemeral merging label

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98543

Addresses https://github.com/pytorch/test-infra/issues/3950

Test Plan: Ran a dry run on this pr. The label showed up while trying to merge
<img width="354" alt="Screenshot 2023-04-06 at 4 57 48 PM" src="https://user-images.githubusercontent.com/13758638/230514276-1ac70b58-d2d1-4e4b-892b-a957bf156063.png">

And then disappeared after failing
<img width="373" alt="Screenshot 2023-04-06 at 5 00 11 PM" src="https://user-images.githubusercontent.com/13758638/230514470-38b15ec7-cfd9-4efe-b6e8-0f9af5577c62.png">

There's also the trail of adding and removing the "merging" label at the bottom

Notes: This is slightly buggy sometimes. For example when the merge failed when I was editing this textbox, the label did not disappear.

PaliC committed:
Add ephemeral merging label

[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"




[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"




[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"




[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

[ghstack-poisoned]

malfet commented on .github/scripts/github_utils.py:
```diff
@@ -127,4 +128,10 @@ def gh_post_commit_comment(
 def gh_delete_comment(org: str, repo: str, comment_id: int) -> None:
     url = f"https://api.github.com/repos/{org}/{repo}/issues/comments/{comment_id}"
-    gh_fetch_url(url, method="DELETE")
+    # gh_fetch_url(url, method="DELETE")
+    print(
+        cast(
+            List[Dict[str, Any]],
+            [],
+        )
+    )
```
Err, is this a draft still?

malfet commented on .github/scripts/label_utils.py:
```diff
@@ -83,6 +83,13 @@ def gh_add_labels(
 
 
+def gh_remove_label(org: str, repo: str, pr_num: int, label: str) -> None:
+    gh_fetch_json(
```
Why not use `gh_fetch_url` here since you do not care about the output anyway?

PaliC replied:
true I was just copying over pattern

malfet reviewed (changes requested):
Looks like this PR still contains some test code. Also, it would be good to add unit-test (or outline a test plan)

I usually test mergebot changes on my personal repo, can add you as collaborator if you want

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

Test Plan: Ran a dry run on this pr. The label showed up while trying to merge
<img width="354" alt="Screenshot 2023-04-06 at 4 57 48 PM" src="https://user-images.githubusercontent.com/13758638/230514276-1ac70b58-d2d1-4e4b-892b-a957bf156063.png">

And then disappeared after failing
<img width="373" alt="Screenshot 2023-04-06 at 5 00 11 PM" src="https://user-images.githubusercontent.com/13758638/230514470-38b15ec7-cfd9-4efe-b6e8-0f9af5577c62.png">

Notes: This is slightly buggy sometimes. For example when the merge failed when I was editing this textbox, the label did not disappear.

[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

Test Plan: Ran a dry run on this pr. The label showed up while trying to merge
<img width="354" alt="Screenshot 2023-04-06 at 4 57 48 PM" src="https://user-images.githubusercontent.com/13758638/230514276-1ac70b58-d2d1-4e4b-892b-a957bf156063.png">

And then disappeared after failing
<img width="373" alt="Screenshot 2023-04-06 at 5 00 11 PM" src="https://user-images.githubusercontent.com/13758638/230514470-38b15ec7-cfd9-4efe-b6e8-0f9af5577c62.png">

Notes: This is slightly buggy sometimes. For example when the merge failed when I was editing this textbox, the label did not disappear.

[ghstack-poisoned]

malfet commented on .github/scripts/github_utils.py:
```diff
@@ -55,6 +55,7 @@ def gh_fetch_json(
     url: str,
     params: Optional[Dict[str, Any]] = None,
     data: Optional[Dict[str, Any]] = None,
+    method: Optional[str] = None,
```
There is nothing wrong with this change, but technically speaking it's no longer needed

malfet commented on .github/scripts/label_utils.py:
```diff
@@ -77,12 +77,21 @@ def gh_get_labels(org: str, repo: str) -> List[str]:
-        f"https://api.github.com/repos/{org}/{repo}/issues/{pr_num}/labels",
+    gh_fetch_url(
+        url=f"https://api.github.com/repos/{org}/{repo}/issues/{pr_num}/labels",
+        headers={"Accept": "application/vnd.github.v3+json"},
```
Again, not sure why this is needed (i.e. only important if one is going to parse the output)
```suggestion
```

malfet commented on .github/scripts/label_utils.py:
```diff
@@ -77,12 +77,21 @@ def gh_get_labels(org: str, repo: str) -> List[str]:
+def gh_remove_label(org: str, repo: str, pr_num: int, label: str) -> None:
+    gh_fetch_url(
+        url=f"https://api.github.com/repos/{org}/{repo}/issues/{pr_num}/labels/{label}",
+        headers={"Accept": "application/vnd.github.v3+json"},
```
Same as above: only important if one is going to parse the output
```suggestion
```

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1983,7 +1992,8 @@ def handle_exception(e: Exception, title: str = "Merge failed") -> None:
         gh_post_pr_comment(org, project, args.pr_num, message, dry_run=args.dry_run)
         return
-
+    if "merging" not in pr.get_labels():
+        pr.add_numbered_label("merging")
```
Please do not hardcode string constants (because in the long run, somebody makes a typo which would be very hard to debug)
Also, if you are checking that label is not there, why not call `gh_add_labels` directly?
```suggestion
    if MERGE_IN_PROGRESS not in pr.get_labels():
        gh_add_label(org, project, args.pr_num, [MERGE_IN_PROGRESS])
```

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -2020,6 +2030,7 @@ def handle_exception(e: Exception, title: str = "Merge failed") -> None:
             )
         else:
             print("Missing comment ID or PR number, couldn't upload to Rockset")
+    pr.remove_label("merging")
```
```suggestion
    pr.remove_label(MERGE_IN_PROGRESS)
```

malfet commented on .github/scripts/trymerge.py:
```diff
@@ -1038,6 +1043,10 @@ def add_numbered_label(self, label_base: str) -> None:
         gh_add_labels(self.org, self.project, self.pr_num, [label])
 
+    def remove_label(self, label: str) -> None:
+        if label in (self.get_labels() if self.get_labels() is not None else []):
+            gh_remove_label(self.org, self.project, self.pr_num, label)
```
Nit
```suggestion
        if self.get_labels() is not None and label in self.get_labels():
            gh_remove_label(self.org, self.project, self.pr_num, label)
```

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

Test Plan: Ran a dry run on this pr. The label showed up while trying to merge
<img width="354" alt="Screenshot 2023-04-06 at 4 57 48 PM" src="https://user-images.githubusercontent.com/13758638/230514276-1ac70b58-d2d1-4e4b-892b-a957bf156063.png">

And then disappeared after failing
<img width="373" alt="Screenshot 2023-04-06 at 5 00 11 PM" src="https://user-images.githubusercontent.com/13758638/230514470-38b15ec7-cfd9-4efe-b6e8-0f9af5577c62.png">

There's also the trail of adding and removing the "merging" label at the bottom

Notes: This is slightly buggy sometimes. For example when the merge failed when I was editing this textbox, the label did not disappear.

[ghstack-poisoned]

PaliC committed:
Update on "Add ephemeral merging label"


Addresses https://github.com/pytorch/test-infra/issues/3950

Test Plan: Ran a dry run on this pr. The label showed up while trying to merge
<img width="354" alt="Screenshot 2023-04-06 at 4 57 48 PM" src="https://user-images.githubusercontent.com/13758638/230514276-1ac70b58-d2d1-4e4b-892b-a957bf156063.png">

And then disappeared after failing
<img width="373" alt="Screenshot 2023-04-06 at 5 00 11 PM" src="https://user-images.githubusercontent.com/13758638/230514470-38b15ec7-cfd9-4efe-b6e8-0f9af5577c62.png">

There's also the trail of adding and removing the "merging" label at the bottom

Notes: This is slightly buggy sometimes. For example when the merge failed when I was editing this textbox, the label did not disappear.

[ghstack-poisoned]

PaliC commented:
@pytorchbot merge
